### ISSUE TITLE
Use vector calling convention on windows

### ISSUE THREAD

**Issue by yuyichao** [shibatch/sleef#245]
After collecting some evidence and doing some research, I believe sleef is not using the vector calling convention on windows.

1. there is no mention of `__vectorcall` in the code. MSVC does **NOT** use the vectorcall calling convention by default.
2. there's no mention of the `@@n` name suffix in the code or in any issues mentioning windows build, e.g. https://github.com/shibatch/sleef/issues/172. The errors in https://github.com/shibatch/sleef/issues/172#issuecomment-366713545 suggests that the library is not compiled to use the vector calling convention.
3. I disassembled one of the few msvc compiled sleef I can find, which is from [the issue mentioned above,](https://github.com/shibatch/sleef/issues/172#issuecomment-369821650), which gives

    ```asm
    Dump of assembler code for function sleef!Sleef_fmaxf8_fma4:
       0x000000018006a450 <+0>:     vmovups (%rdx),%ymm2
       0x000000018006a454 <+4>:     vmovups (%rcx),%ymm4
       0x000000018006a458 <+8>:     vcmpneqps %ymm2,%ymm2,%ymm3
       0x000000018006a45d <+13>:    vcmpgt_oqps %ymm2,%ymm4,%ymm1
       0x000000018006a462 <+18>:    vblendvps %ymm1,%ymm4,%ymm2,%ymm2
       0x000000018006a468 <+24>:    vblendvps %ymm3,%ymm4,%ymm2,%ymm0
       0x000000018006a46e <+30>:    retq
       0x000000018006a46f <+31>:    int3
    End of assembler dump.
    ```

The two arguments are passed by reference (the first two instructions loads them) and returned in `ymm0`. IIRC this is what clang does without `__vectorcall` (I don't have MSVC to test) so I believe this is just using the non-vectorcall ABI.

I'm not sure what the intention (though I believe I've seen mentioning of the vector ABI so I thought the intention is to use it) so here is just a dump of the situation on windows that I collected which can hopefully make it clear why I needed the changes in my recent PRs and to avoid duplicated work since I'm not sure if I'll work on making my hacky version generally usable.

1. GCC
    * [Does not support __vectorcall](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=89485)
    * [Has stack alignment issue with `__m256` arguments](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=54412)
    * The non vectorcall ABI appears to be different from that from clang for 256bit vector (GCC uses sret for the return value whereas clang returns the vector in the register)
2. Clang
    * I assume it is compatible with MSVC (I don't have MSVC to check)
    * Support `__vectorcall`
    * `__vectorcall` does not "work" for 512bit vectors (they still get passed on the stack). From various random sources it seems that this might be the case for MSVC as well.
    * Not really related to ABI but I found clang usually generate worse code for these.... Potentially related to register zeroing.

So it seems that in any case there might not be a way to compile a binary that works for both gcc and clang (and MSVC) unless one pass the argument and return values by reference explicitly. In this case it might be OK to just require the vectorcall ABI at least for 256bit vectors since that should work for both MSVC and clang and is more efficient.

**Comment by carlkl** [comment]
Maybe this flag helps for MSVC builds:
`Set the default calling convention to __vectorcall (/Gv)`

**Comment by yuyichao** [comment]
Ideally non vector function shouldn't use the vector calling convention

**Comment by carlkl** [comment]
@yuyichao, 

concerning GCC's calling convention as used by the mingw-w64 toolchain take a look at: https://sourceforge.net/p/mingw-w64/wiki2/MinGW%20x64%20Software%20convention/
The vectorcall calling convention is not supported, but the default calling conventions for C libraries are compatible with MSVC.

Concerning the AVX bug see also:
https://developercommunity.visualstudio.com/content/problem/19160/regression-from-vs-2015-in-ssseavx-instructions-ge.html and https://software.intel.com/en-us/forums/intel-isa-extensions/topic/279587
Unfortunately gcc is using aligned instructions (VMOVAPS, VMOVAPD) instead of unaligned instructions (VMOVUPS, VMOVUPD). Since NEHALEM the unaligned instructions have more or less the same performance as the aligned instructions. 
See also: https://github.com/msys2/MSYS2-packages/issues/1209#issuecomment-379576367

**Comment by yuyichao** [comment]
> The vectorcall calling convention is not supported, but the default calling conventions for C libraries are compatible with MSVC.

I don't believe this is the case. Either this or clang is not compatible with MSVC and from the disassembly I pasted above I don't believe that is the case. The argument passing is indeed consistent between gcc and MSVC but the result isn't. MSVC/Clang seems to always pass the result in the register whereas GCC pass the result by reference.

> Concerning the AVX bug see also:

Errr, I don't think any of those are really related to what sleef should do? I think pretty much the only relevant info is that the only way to support GCC on windows right now is to pass 256bit vectors by reference.

> Since NEHALEM the unaligned instructions have more or less the same performance as the aligned instructions.

When the data is aligned, I'm pretty sure there can still be a performance penalty if the pointer is not aligned. The issue is that GCC failed to aligned the stack, which it knows how to do but just didn't do it, so it should be possible for it to align the stack which will have performance benefit. Once the stack is aligned though, it doesn't hurt to keep using aligned instructions. Obviously using unaligned instruction is a way to workaround that issue (though that's completely Off topic for this unless gcc has a flag for that...) doing that won't be the best solution performance wise.

**Comment by yuyichao** [comment]
And I would also summarize what I had above along a different dimension. Ideally, on windows, the support for different functions should be considered separately, even though doing that is harder on the coding side...

1. scaler functions:

    Ideally this should not use `__vectorcall`. Adding such attribute will not change anything WRT the calling convention but will make GCC complain and will make it unable to find the symbol (The symbol will still be renamed with the `@@n` suffix which the GCC won't understand)

2. 128bit functions.

    This is the one that need some tradeoff. Without `__vectorcall`, GCC, clang and MSVC agrees on the calling convention. This needs to come without `__vectorcall` (and with the performance penalty) if GCC is to be supported with the same binary. In principle, when compiled with clang/MSVC, one could generate two versions of the same function since their symbol do not conflict. This will be the ideal solution but I'm not sure if it's worth doing...

3. 256bit

    The only calling convention GCC uses for this isn't consistent with anyone else so there's no point in supporting it (it could also mean that since the GCC calling convention is buggy they might be able to fix it to be consistent with clang and MSVC without breaking anyone's code...).

    This means that MSVC and Clang could and probably should use `__vectorcall`.

4. 512bit

    Same as 256bit for GCC, it doesn't use the same calling convention so supporting GCC isn't possible. MSVC and Clang doesn't do anything special for `__vectorcall` though so either way works.

Other than these, the additional consideration is that if we add `__vectorcall`, the alias code needs to be tweaked to handle the symbol renaming...

**Comment by thebrightestlights** [comment]
@yuyichao,
About GCC, at the bottom of the page you linked (https://gcc.gnu.org/bugzilla/show_bug.cgi?id=54412) it is mentioned that the assembler option ``-muse-unaligned-vector-move`` (which can be given to gcc via ``-Wa``) can be used to encode all vector move instructions as unaligned. After building sleef with this option on gcc (via mingw-w64) it was possible to run Sleef on my machine. So I think it is still possible to support GCC on Windows for 256bit functions by means of this flag, which could be added as a configuration.

**Comment by thebrightestlights** [comment]
Although, I forgot to mention, the caller must also compile (at least the calling parts) to use unaligned moves, so it might not be worth it, or only as a short remark in the documentation
